File: scala-style/src/router.scala
/**
 * Created with IntelliJ IDEA.
 * User: Stig
 * Date: 10/17/12
 * Time: 7:00 PM
 * To change this template use File | Settings | File Templates.
 */

import java.io._
import java.net._

object router {

  def main(args: Array[String]) = {
    require(args.length < 6, "Must have 5 or less arguments")
    //TODO: Test where things are failing and why
    //TODO: Maybe try adding in some exceptions here
    println("check1")
    val servSock = new ServerSocket(2222)
    val clientSock = servSock.accept()
    val ipList = buildRoutingTable(args, List(), 0)
    var timesRun = 0

    while(true){
      println("This program has been consecuatively run " + timesRun)
      println("Waiting to receive message...")
      //Get (message, (ipOfServer, port), elapsed time of routing table lookup)
      val messageAndIp = buildData(clientSock, ipList)
      //break apart messageAndIp into it's parts
      val routingTableLookupTime = messageAndIp.last.toLong
      val toServer = new Socket(messageAndIp.tail.head.toString, messageAndIp.init.last.toInt)

      sendMessage(messageAndIp.head.getBytes, toServer)
      println("Message sent...\nTransporting data...")

      transportData(clientSock, toServer)

      println("Data has been transported")

      servSock.close; clientSock.close;

      timesRun+=1
      val avgLookup = routingTableLookupTime/timesRun.toLong
      println("Current routing table lookup time " + routingTableLookupTime + "ms")
      println("Average routing table lookup time " + avgLookup + "ms")
    }
  }

  def buildRoutingTable(args: Array[String], list: List[(String, String)], count: Int): List[(String, String)] = {
    if (args.isEmpty) list
    else buildRoutingTable(args, list ++ List((args(count), "2222")), count + 1)
  }

  def buildData(clientSock: Socket, ipList: List[(String, String)]): List[String] = {
    def receiveMessage(): List[String] = {
      val receiveData: Array[Byte] = null
      val inFromClient = new DataInputStream(clientSock.getInputStream)
      inFromClient.read(receiveData)
      var message = new String(receiveData)
      message = message.trim
      val messages = message.split(":")
      messages.toList
    }

    def searchIpList(serverIP: String, ipList: List[(String, String)]): (String, String) = {
      if (serverIP == ipList.head._1) ipList.head
      else if (ipList.isEmpty)
        throw sys.error("There is no server on the network that matches your request")
      else searchIpList(serverIP, ipList.tail)
    }

    val messages = receiveMessage()
    val start = System.currentTimeMillis()
    val ipToSendTo = searchIpList(messages(2), ipList)
    val elapsedTime = System.currentTimeMillis()-start
    List(messages(0), ipToSendTo._1, ipToSendTo._2, elapsedTime.toString)
  }

  def sendMessage(sendData: Array[Byte], toServer: Socket) = {
    val out = new DataOutputStream(toServer.getOutputStream)
    out.write(sendData)
  }

  def transportData(clientSock: Socket, fromServer: Socket) = {
    val in = fromServer.getInputStream
    val out = clientSock.getOutputStream
    val receiveFile: Array[Byte] = null
    var bytesRead = in.read(receiveFile)
    while(bytesRead != -1) {
      out.write(receiveFile, 0, bytesRead)
      bytesRead = in.read(receiveFile)
    }
  }
}

File: part-2/src/Router.java
import java.io.*;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.LinkedList;
import java.util.Scanner;

/**
 * Created with IntelliJ IDEA.
 * User: Stig
 * Date: 10/31/12
 * Time: 7:00 PM
 * To change this template use File | Settings | File Templates.
 */
public class Router {

    //For ip list
    private static LinkedList<String> ipList = new LinkedList<String>();
    private static String ip;
    private static String otherRouterIP;

    //Sockets
    private static ServerSocket serverSocket;
    private static Socket client, otherRouter;

    //Streams
    private static InputStream inFromClient, inFromOtherRouter;
    private static OutputStream outToClient, outToOtherRouter;

    //Data buffers
    static byte[] sendData = new byte[64000];
    static byte[] receiveData = new byte[64000];

    String command;
    static int bytesRead;
    private static Scanner scan = new Scanner(System.in);

    public Router(){
        try {
            serverSocket = new ServerSocket(2222);
        } catch (IOException e) {
            e.printStackTrace();  //To change body of catch statement use File | Settings | File Templates.
        }
    }

    public static void main(String[] args){
        System.out.println("Please enter all IPs of all clients on this side" +
        "\nType the letter <e> when finished");

        try {
            while(!(ip = scan.next()).equals("e")){
                ipList.add(ip);
            }
        } catch (Exception e) {
            System.out.println("Error adding IP");  //To change body of catch statement use File | Settings | File Templates.
        }

        System.out.println("Please enter the IP of the other router.");

        otherRouterIP = scan.next();

        Router router = new Router();

        System.out.println("Waiting to receive command...");
        try {
            client = serverSocket.accept();
        } catch (IOException e) {
            e.printStackTrace();  //To change body of catch statement use File | Settings | File Templates.
        }

        router.receiveMessage(router.receiveCommand());
    }

    private String receiveCommand(){
        try {
            inFromClient = new DataInputStream(client.getInputStream());
            inFromClient.read(receiveData);
        } catch (IOException e) {
            e.printStackTrace();  //To change body of catch statement use File | Settings | File Templates.
        }
        command = new String(receiveData);
        return command.trim();
    }

    private void receiveMessage(String command){
            if(command.equals("getIPs")){
                getIPs();
            }
            else if (command.equals("sendIPs")){
                sendIPs();
            }
    }

    private static void getIPs(){
        System.out.println("Grabbing IPs...");
        try {
            otherRouter = new Socket(otherRouterIP, 2222);
            inFromOtherRouter = new DataInputStream(otherRouter.getInputStream());
            outToOtherRouter = new DataOutputStream(otherRouter.getOutputStream());
            outToClient = new DataOutputStream(client.getOutputStream());

            outToOtherRouter.write("sendIPs".getBytes());

            receiveData = new byte[64000];
            inFromOtherRouter.read(receiveData);
            outToClient.write(receiveData);

            while( (bytesRead = inFromOtherRouter.read(receiveData)) != -1){
                System.out.println(new String(receiveData).trim());
                outToClient.write(receiveData, 0, bytesRead);
            }

            inFromOtherRouter.close();
            outToOtherRouter.close();
            outToClient.close();
            otherRouter.close();
        } catch (IOException e) {
            e.printStackTrace();  //To change body of catch statement use File | Settings | File Templates.
        }
    }

    private static void sendIPs(){
        System.out.println("Sending IPs...");
        try {
            inFromOtherRouter = new DataInputStream(client.getInputStream());
            outToOtherRouter = new DataOutputStream(client.getOutputStream());
        } catch (IOException e) {
            e.printStackTrace();  //To change body of catch statement use File | Settings | File Templates.
        }

        String ipListSize = String.valueOf(ipList.size());
        try {
            outToOtherRouter.write(ipListSize.getBytes());
        } catch (IOException e) {
            e.printStackTrace();  //To change body of catch statement use File | Settings | File Templates.
        }

        System.out.println("passed streams");
        try {
            for(int i=0; i<ipList.size(); i++){
                System.out.println(ipList.get(i));
                sendData = ipList.get(i).getBytes();
                outToOtherRouter.write(sendData);
            }
        } catch (IOException e) {
            e.printStackTrace();  //To change body of catch statement use File | Settings | File Templates.
        } catch (NullPointerException e){
            System.out.println("Cannot send IPs...");
        }
    }
}
